core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -977,4 +980,35 @@ class KafkaApis(val requestChannel: RequestChannel,
     if (!authorize(request.session, ClusterAction, Resource.ClusterResource))
       throw new ClusterAuthorizationException(s"Request $request is not authorized.")
   }
+
+  def getProtocolVersions(): java.util.List[ProtocolVersion] = {
+    val protocolVersions: java.util.List[ProtocolVersion] = new java.util.ArrayList[ProtocolVersion]()
+    ApiKeys.values().foreach(apiKey => {
+      var index: JShort = JShort.valueOf("-1")
+      val reqs = Protocol.REQUESTS(apiKey.id).flatMap(r => {
+        index = (index + 1).toShort
+        if (r != null)
+          Some(index)
+        else
+          None
+      })
+
+      protocolVersions.add(new ProtocolVersion(apiKey.id, apiKey.name, reqs.toSeq.asJava, Protocol.DEPRECATED_VERSIONS(apiKey.id).toSeq.asJava.asInstanceOf[java.util.List[JShort]]))
+    })
+    protocolVersions
+  }
+
+  def handleProtocolVersionRequest(request: RequestChannel.Request) {
+    val responseHeader = new ResponseHeader(request.header.correlationId)
+    val responseBody = if (!authorize(request.session, Describe, Resource.ClusterResource)) {
+      ProtocolVersionResponse.fromError(Errors.CLUSTER_AUTHORIZATION_FAILED)
+    } else {
+      new ProtocolVersionResponse(Errors.NONE.code(), getProtocolVersions())
+    }
+    requestChannel.sendResponse(new RequestChannel.Response(request, new ResponseSend(request.connectionId, responseHeader, responseBody)))
+  }
+
+  def handleUnknownRequest(request: RequestChannel.Request) {
```

Comment:
This is more based on discussions on KIP-35, quoting from wiki.

> If a request is received for an unknown protocol request type, or for an unsupported version of a known type, the broker should respond with an empty response only including the common Length+CorrId header with Length=0, instead of closing the connection. The effect of this empty message received by the client is that the client will fail to parse the response (since it will expect a proper non-empty reply) and throw an error to the application with the correct context - that of a protocol parsing failure for request type XYZ. All existing clients should be able to handle this gracefully without any alterations to the code, while updated clients supporting the proposals in this KIP can handle the empty response accordingly by returning a "Request not supported by server" error the application. The level of detail in the returned error message to the application naturally varies between client implementations but is still by far better than a closed connection which says nothing of the underlying error.